cuda/test/stop/residual_norm_kernels.cpp

```diff
@@ -0,0 +1,366 @@
+    std::shared_ptr<gko::ReferenceExecutor> ref_;
+};
+
+
+TEST_F(RelativeResidualNorm, WaitsTillResidualGoal)
+{
+    auto res = gko::initialize<Mtx>({100.0}, ref_);
+    auto d_res = Mtx::create(cuda_);
+    d_res->copy_from(res.get());
+    std::shared_ptr<gko::LinOp> rhs = gko::initialize<Mtx>({10.0}, ref_);
+    std::shared_ptr<gko::LinOp> d_rhs = Mtx::create(cuda_);
+    d_rhs->copy_from(rhs.get());
+    auto criterion = factory_->generate(nullptr, d_rhs, nullptr, d_res.get());
+    bool one_changed{};
+    constexpr gko::uint8 RelativeStoppingId{1};
+    gko::Array<gko::stopping_status> stop_status(ref_, 1);
+    stop_status.get_data()[0].reset();
+    stop_status.set_executor(cuda_);
+
+    ASSERT_FALSE(
+        criterion->update()
+            .residual_norm(d_res.get())
+            .check(RelativeStoppingId, true, &stop_status, &one_changed));
+
+    res->at(0) = tol * 1.0e+1;
+    d_res->copy_from(res.get());
+    ASSERT_FALSE(
+        criterion->update()
+            .residual_norm(d_res.get())
+            .check(RelativeStoppingId, true, &stop_status, &one_changed));
+    stop_status.set_executor(ref_);
+    ASSERT_FALSE(stop_status.get_data()[0].has_converged());
+    stop_status.set_executor(cuda_);
+    ASSERT_FALSE(one_changed);
+
+    res->at(0) = tol;
+    d_res->copy_from(res.get());
+    ASSERT_TRUE(
+        criterion->update()
+            .residual_norm(d_res.get())
+            .check(RelativeStoppingId, true, &stop_status, &one_changed));
+    stop_status.set_executor(ref_);
+    ASSERT_TRUE(stop_status.get_data()[0].has_converged());
+    ASSERT_TRUE(one_changed);
+}
+
+
+TEST_F(RelativeResidualNorm, WaitsTillResidualGoalMultipleRHS)
+{
+    auto res = gko::initialize<Mtx>({100.0, 100.0}, ref_);
+    auto d_res = Mtx::create(cuda_);
+    d_res->copy_from(res.get());
+    std::shared_ptr<gko::LinOp> rhs = gko::initialize<Mtx>({10.0, 10.0}, ref_);
+    std::shared_ptr<gko::LinOp> d_rhs = Mtx::create(cuda_);
+    d_rhs->copy_from(rhs.get());
+    auto criterion = factory_->generate(nullptr, d_rhs, nullptr, d_res.get());
+    bool one_changed{};
+    constexpr gko::uint8 RelativeStoppingId{1};
+    gko::Array<gko::stopping_status> stop_status(ref_, 2);
+    stop_status.get_data()[0].reset();
+    stop_status.get_data()[1].reset();
+    stop_status.set_executor(cuda_);
+
+    ASSERT_FALSE(
+        criterion->update()
+            .residual_norm(d_res.get())
+            .check(RelativeStoppingId, true, &stop_status, &one_changed));
+
+    res->at(0, 0) = tol * 1.0e+1;
+    d_res->copy_from(res.get());
+    ASSERT_FALSE(
+        criterion->update()
+            .residual_norm(d_res.get())
+            .check(RelativeStoppingId, true, &stop_status, &one_changed));
+    stop_status.set_executor(ref_);
+    ASSERT_TRUE(stop_status.get_data()[0].has_converged());
+    stop_status.set_executor(cuda_);
+    ASSERT_TRUE(one_changed);
+
+    res->at(0, 1) = tol * 1.0e+1;
+    d_res->copy_from(res.get());
+    ASSERT_TRUE(
```

Comment:
In the previous test, WaitsTillResidualGoal, `res` needs to be `tol` to satisfy the stopping criterion, but `res` in this test only needs to `tol * 10`. Do I misunderstand somthing?

Reply:
I got these wrong before, my idea was to start with the same initial residual and right-hand side for all criteria and then check the different behavior of the criteria:
- ResidualReduction is satisfied when `|r| / |r_0| < tol`, so when `|r| < 100 * tol`
- RelativeResidualNorm is staisfied when `|r| / |b| < tol`, so when `|r| < 10 * tol`
- AbsoluteResidualNorm is satisfied when `|r| < tol`